source/common/http/conn_pool_base.cc

```diff
@@ -1,40 +1,292 @@
+  client.state_ = new_state;
+  if (&old_list != &new_list) {
+    client.moveBetweenLists(old_list, new_list);
+  }
 }
 
-void ConnPoolImplBase::ActiveClient::recordConnectionSetup() {
-  conn_connect_ms_->complete();
-  conn_connect_ms_.reset();
+void ConnPoolImplBase::closeIdleConnections() {
+  // Create a separate list of elements to close to avoid mutate-while-iterating problems.
+  std::list<ActiveClient*> toClose;
+
+  // Possibly-idle connections are always in the ready_clients_ list
+  for (auto& client : ready_clients_) {
+    if (!client->hasActiveRequests()) {
+      toClose.push_back(client.get());
+    }
+  }
+
+  for (auto& entry : toClose) {
+    entry->close();
+  }
 }
 
-void ConnPoolImplBase::ActiveClient::disarmConnectTimeout() {
-  if (connect_timer_) {
-    connect_timer_->disableTimer();
-    connect_timer_.reset();
+void ConnPoolImplBase::drainConnections() {
+  closeIdleConnections();
+
+  // closeIdleConnections() closes all connections in ready_clients_ with no active requests,
+  // so all remaining entries in ready_clients_ are serving requests. Move them and all entries
+  // in busy_clients_ to draining.
+  while (!ready_clients_.empty()) {
+    setActiveClientState(*ready_clients_.front(), ActiveClient::State::DRAINING);
+  }
+
+  // Changing busy_clients_ to DRAINING does not move them between lists,
+  // so use a for-loop since the list is not mutated.
+  ASSERT(&owningList(ActiveClient::State::DRAINING) == &busy_clients_);
+  for (auto& busy_client : busy_clients_) {
+    setActiveClientState(*busy_client, ActiveClient::State::DRAINING);
   }
 }
 
-ConnPoolImplBase::ActiveClient::ConnectionState ConnPoolImplBase::ActiveClient::connectionState() {
-  // We don't track any failure state, as the client should be deferred destroyed once a failure
-  // event is handled.
-  if (connect_timer_) {
-    return Connecting;
+void ConnPoolImplBase::checkForDrained() {
+  if (drained_callbacks_.empty()) {
+    return;
   }
 
-  return Connected;
+  closeIdleConnections();
+
+  if (pending_requests_.empty() && ready_clients_.empty() && busy_clients_.empty()) {
+    ENVOY_LOG(debug, "invoking drained callbacks");
+    for (const DrainedCb& cb : drained_callbacks_) {
+      cb();
+    }
+  }
+}
+
+void ConnPoolImplBase::onConnectionEvent(ConnPoolImplBase::ActiveClient& client,
+                                         Network::ConnectionEvent event) {
+  if (event == Network::ConnectionEvent::RemoteClose ||
+      event == Network::ConnectionEvent::LocalClose) {
+    // The client died.
+    ENVOY_CONN_LOG(debug, "client disconnected, failure reason: {}", *client.codec_client_,
+                   client.codec_client_->connectionFailureReason());
+
+    Envoy::Upstream::reportUpstreamCxDestroy(host_, event);
+    if (client.closingWithIncompleteRequest()) {
+      Envoy::Upstream::reportUpstreamCxDestroyActiveRequest(host_, event);
+    }
+
+    const bool check_for_drained = client.closingWithIncompleteRequest();
```

Comment:
you're calling `closingWithIncompleteRequest` in the previous if block as well, should probably consolidate?